### ISSUE TITLE
GTM failover，after reconnect gtm_proxy all， Sometime gtm_proxy get coredump

### ISSUE THREAD

**Issue by xuegang** [postgres-x2/postgres-x2#135]
Test process：
 1.Perform 20 concurrent insert in coordinator
 2.kill -9 gtm
 3.failover gtm
 4.reconnect gtm_proxy all ----  some time gtm_proxy get coredump，

Stack ：
Core was generated by `/opt/rdb3.1_install/bin/gtm_proxy -D /opt/rdb3.1_data/nodes/gtm_pxy'.
Program terminated with signal 11, Segmentation fault.
#0 0x000000000041fd79 in Recovery_PGXCNodeDisconnect (myport=0x0) at register_common.c:727

727 GTM_PGXCNodeType type = myport->remote_type;
Missing separate debuginfos, use: debuginfo-install glibc-2.17-55.el7.x86_64
(gdb)#1 0x0000000000425b0d in G bt
#0 0x000000000041fd79 in Recovery_PGXCNodeDisconnect (myport=0x0) at register_common.c:727

TMProxy_HandleDisconnect (conninfo=0x1327658, gtm_conn=0x7fd07402d110) at proxy_main.c:2664
#2 0x00000000004268c0 in GTMProxy_ProcessPendingCommands (thrinfo=0x1337a70) at proxy_main.c:2901
#3 0x0000000000423229 in GTMProxy_ThreadMain (argp=0x1337a70) at proxy_main.c:1510
#4 0x00000000004281bb in GTMProxy_ThreadMainWrapper (argp=0x1337a70) at proxy_thread.c:319
#5 0x00007fd07c2addf3 in start_thread () from /lib64/libpthread.so.0
#6 0x00007fd07bfdb3dd in clone () from /lib64/libc.so.6

163 gtm_proxy log：
1:140533409138432:2014-09-03 20:39:31.061 CST -LOG: cleaning up client disconnection
LOCATION: GTMProxy_ProcessPendingCommands, proxy_main.c:2899
1:140533409138432:2014-09-03 20:39:37.067 CST -LOG: cleaning up client disconnection
LOCATION: GTMProxy_ProcessPendingCommands, proxy_main.c:2899
1:140533417408320:2014-09-03 20:39:50.626 CST -LOG: I'm the main thread. Accepted SIGUSR1.
LOCATION: GTMProxy_SigleHandler, proxy_main.c:488
1:140533417408320:2014-09-03 20:39:50.626 CST -LOG: Main Thread reconnecting to new GTM.
LOCATION: ServerLoop, proxy_main.c:1029
1:140533417408320:2014-09-03 20:39:50.627 CST -LOG: Reconnected.
LOCATION: ServerLoop, proxy_main.c:1031
1:140533409138432:2014-09-03 20:39:50.628 CST -LOG: cleaning up client disconnection
LOCATION: GTMProxy_ProcessPendingCommands, proxy_main.c:2899

**Comment by koichi-szk** [comment]
Okay, I'll write some tentative fix and post as pull request.   Could you test it and see if it works?  Maybe late today or tomorrow afternoon.

Regards;
Koichi

**Comment by xuegang** [comment]
We have fixed this problem. Later I will send pull request and analyze process to you.

**Comment by xuegang** [comment]
Signal(from reconnect gtm_proxy all ) processing disrupt the normal process of gtm proxy work thread. So get coredump

Analyze process as follows:
 In GTMProxy_ThreadMain(void *argp)  
        for (;;)
        {
           1.setjmp_again: 
           if (sigsetjmp(thrinfo->longjmp_env, 1) == 0)
           { Disable_Longjmp(); }

```
     2.GTMProxy_ProcessPendingCommands //set conninfo->con_port = NULL; //At the point where the problem occurs

       ...

     3.GTMProxy_ThreadRemoveConnection //remomve connection 

  }
```

  when work thread execution between 2 and 3. Do some something as follows:
      1、kill -9 gtm
      2、failover gtm
      3、reconnect gtm_proxy all  
        3.1 this command will send siguser1 to gtm proxy，main thread run signal process function: GTMProxy_SigleHandler，then send siguser2 to work thread.
        3.2 work thread run GTMProxy_SigleHandler and forced to jump to "1.setjmp_again":
        3.3 work thread again to "2:GTMProxy_ProcessPendingCommands"，because conninfo->con_port = NULL，so got gtm proxy coredump.

**Comment by koichi-szk** [comment]
Okay, now we have three pull requests related to the issue.

Xuegang, could you summarize your progress until now what commits are needed, all or some of them?  Fix issue #135 looks very simple and seems to work.  The change is in `GTMProxy_HandleDisconnect()` so it is reasonable just to return if `conninfo->con_port == NULL`.

I will push this patch and close the issue.

Thank you.

**Comment by koichi-szk** [comment]
Committed.  Will close this thread.
